[FLINK-12732][state-processor-api] Add savepoint reader for consuming partitioned operator state

sjwiesman opened the pull request:
## What is the purpose of the change

This is the second PR for FLIP-43 adding the functionality to read partitioned operator state from a state snapshot. It is based on #8615 and only the final 4 commits should be reviewed.

## Brief change log

62dd050 - Some basic refactoring, this pulls the `NeverCompleteFuture` into a reusable utility

38eaa66 - Implements a shim `Environment` on-top of the available user `RuntimeContext`

9689001 - Adds the new functionality to read keyed state out of a snapshot

be0552c - Adds new documentation

## Verifying this change

*(Please pick either of the following options)*

This change added tests and can be verified as follows:
UT and IT tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

sjwiesman commented:
@flinkbot attention @tzulitai

tzulitai commented:
@flinkbot approve description
@flinkbot approve consensus

tzulitai commented on flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/functions/KeyedStateReaderFunction.java:
```diff
@@ -0,0 +1,82 @@
+	 * Initialization method for the function. It is called before {@link #readKey(Object,
+	 * Context, Collector)} and thus suitable for one time setup work.
+	 *
+	 * <p>This is the only method that my register state descriptors within a {@code
```
typo: only method that `may` ...

tzulitai commented on flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/BufferingCollector.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Queue;
+
+/**
+ * A collector that buffers elements in memory to be pulled latter by downstream operators.
```
typo: `latter` -> `later`

tzulitai commented on flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/KeyedStateInputFormat.java:
```diff
@@ -0,0 +1,253 @@
+		final DefaultKeyedStateStore keyedStateStore = new DefaultKeyedStateStore(keyedStateBackend, getRuntimeContext().getExecutionConfig());
+
+		final List<StateDescriptor<?, ?>> stateDescriptors;
+		try (SavepointRuntimeContext ctx = new SavepointRuntimeContext(getRuntimeContext(), keyedStateStore)) {
```
Please see my comment concerning whether `SavepointRuntimeContext` should really be an `AutoClosable`.

tzulitai replied:
I find something like:

```
FunctionUtils.openFunction(userFunction, new Configuration());
ctx.disableStateRegistration();
```

much more understandable here.

tzulitai commented on flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/KeyedStateInputFormat.java:
```diff
@@ -0,0 +1,253 @@
+
+		final List<StateDescriptor<?, ?>> stateDescriptors;
+		try (SavepointRuntimeContext ctx = new SavepointRuntimeContext(getRuntimeContext(), keyedStateStore)) {
+			FunctionUtils.setFunctionRuntimeContext(userFunction, ctx);
```
I feel like we should be setting the runtime context already in the `open()` method, instead of here, since the concern of this method should really only be about getting the key iterator.

tzulitai commented on flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/runtime/VoidTriggerable.java:
```diff
@@ -0,0 +1,43 @@
+ * @param <K> Type of the keys to which timers are scoped.
+ * @param <N> Type of the namespace to which timers are scoped.
+ */
+public final class VoidTriggerable<K, N> implements Triggerable<K, N> {
```
Add `@Internal`.

sjwiesman commented:
Thanks for the review @tzulitai !

I've addressed all the comments including moving the savepoint loader to run once on the client. This is rebased over the fixes on #8615

tzulitai commented:
@flinkbot approve architecture
@flinkbot approve quality

tzulitai commented:
Merging ...
